src/test/basics/base_uint_test.cpp

```diff
@@ -102,9 +126,56 @@ struct base_uint_test : beast::unit_test::suite
         n = beast::zero;
         BEAST_EXPECT(n == z);
 
-        auto hash = sha512Half(u, v, z, n);
-        BEAST_EXPECT(to_string(hash) ==
-            "55CAB39C72F2572057114B4732210605AA22C6E89243FBB5F9943130D813F902");
+        test96 x { (--test96 { z }) ^ (++test96 { z }) };
+        uset.insert(x);
+        BEAST_EXPECT(to_string(x) == "FFFFFFFFFFFFFFFFFFFFFFFE");
+
+        BEAST_EXPECT(uset.size() == 4);
+
+        // SetHex tests...
+        test96 fromHex;
+        BEAST_EXPECT(fromHex.SetHexExact(to_string(u)));
+        BEAST_EXPECT(fromHex == u);
+        fromHex = z;
+
+        // fails with extra char
+        BEAST_EXPECT(! fromHex.SetHexExact("A" + to_string(u)));
+        BEAST_EXPECT(fromHex != u);
+        fromHex = z;
+
+        // fails with extra char at end, but the value is still parsed (?)
+        BEAST_EXPECT(! fromHex.SetHexExact(to_string(u) + "A"));
+        BEAST_EXPECT(fromHex == u);
+        fromHex = z;
+
+        BEAST_EXPECT(fromHex.SetHex(to_string(u)));
+        BEAST_EXPECT(fromHex == u);
+        fromHex = z;
+
+        // leading space/0x allowed if not strict
+        BEAST_EXPECT(fromHex.SetHex("  0x" + to_string(u)));
+        BEAST_EXPECT(fromHex == u);
+        fromHex = z;
+
+        // other leading chars also allowed (ignored)
+        BEAST_EXPECT(fromHex.SetHex("FEFEFE" + to_string(u)));
+        BEAST_EXPECT(fromHex == u);
+        fromHex = z;
+
+        // invalid hex chars should fail (0 replaced with Z here)
+        BEAST_EXPECT(! fromHex.SetHex(
+            boost::algorithm::replace_all_copy(to_string(u), "0", "Z")));
+        BEAST_EXPECT(fromHex != u);
+        fromHex = z;
+
+        BEAST_EXPECT(fromHex.SetHex(to_string(u), true));
+        BEAST_EXPECT(fromHex == u);
+        fromHex = z;
+
+        // strict mode fails with leading chars
+        BEAST_EXPECT(! fromHex.SetHex("  0x" + to_string(u), true));
```

Comment:
@mwshortt You need to understand the history of the code. Someone needed to convert a hex string to a `base_uint`. Then someone else needed to do the same thing with somewhat different rules. And so on, and so forth. The code grew by accretion rather than by design. `SetHexExact` is among the oldest functions in rippled.